poc/ws_client.py

```diff
@@ -0,0 +1,170 @@
+
+        Args:
+            uri (str): A websocket URI to connect to
+
+        Returns:
+            ClientConnection: A new instance of a client connection
+        """
+        try:
+            connection = await connect(uri)
+            return cls(connection, *args, **kwargs)
+        except ConnectionError:
+            # Failed to connect to the server
+            print("Connection failed")
+
+    async def producer_handler(self, websocket):
+        """Waits for new data from client and sends it over the websocket connection
+
+        Args:
+            websocket (websockets.WebSocketProtocol): websocket connection
+        """
+        while True:
+            message = await self.producer()
+            await websocket.send(message)
+
+    async def consumer_handler(self, websocket):
+        """Looks for data from the websocket server, and forwards it to the consumer
+
+        Args:
+            websocket (websockets.WebSocketProtocol): websocket protocol
+        """
+        while True:
+            data = await websocket.recv()
+            print("From Consumer Handler", data)
+            await self.consumer(data)
+
+    async def handler(self):
+        """
+        Handles websocket interactions at a high level
+
+            - A good place for error catching and creating other async tasks
+        """
+        print("Running handler")
+
+        while True:
+            await asyncio.gather(
+                self.consumer_handler(self.ws_connection),
+                self.producer_handler(self.ws_connection),
+            )
+
+
+URI = "wss://demo.piesocket.com/v3/channel_1?api_key=VCXCEuvhGcBDP7XhiJJUDvR1e1D3eiVjgZ9VRiaV&notify_self"
+URI2 = "ws://localhost:8000/ws/20"
+
+
+async def start_simple_async_client():
+    """A simple CLI Chat Client"""
+    ex_client = ExampleClient()
+    client_connection = await ClientConnection.start_connection(
+        URI2, producer=ex_client.producer, consumer=ex_client.consumer
+    )
+
+    # listen for blocking input in the background
+    await client_connection.handler()
+
+
+def _asyncio_thread(loop, task, task_args=[], task_kwargs={}):
+    loop.create_task(task(*task_args, **task_kwargs))
+    loop.run_forever()
+
+
+def start_gui_client():
+    """
+    Complex Chat Client
+
+        - Tkinter eventloop
+        - asyncio eventloop running in a background thread
+    """
+    async_loop = asyncio.get_event_loop()
+    client_gui = ChatClientGUI(async_loop)
+    client_connection = async_loop.run_until_complete(
+        ClientConnection.start_connection(
+            URI2, producer=client_gui.send_message, consumer=client_gui.receive_message
+        )
+    )
+
+    # Start Websocket Thread
+    threading.Thread(
+        target=_asyncio_thread,
+        args=(
+            async_loop,
+            client_connection.handler,
+        ),
+        daemon=True,
+    ).start()
+    client_gui.start_mainloop()
+
+
+"""
+SIMPLE CLIENT
```

Comment:
Why are your single-line docstrings on 3 lines, the [PEP 257](https://peps.python.org/pep-0257/) conventions say:

```py
"""single line summary"""
```

```py
"""single line summary

optional larger
description
"""
```

Reply:
They're serving as headers/comments here, not necessarily docstrings.

Reply:
But still, why do they need 3 entire lines for a one line comment?